Exposes statsd format and send functions

bradlangel opened the pull request:
Motivation for this commit is to give the user access to statsd message
formatting functions and a low level send function in order allow the
user to extend the statsd protocol. Based on top of #5

brendanlong reviewed (commented):
Since this is a pull request, can you make whitespace changes in a different commit, to make it easier to review?

brendanlong commented on statsd_client.ml:
```diff
@@ -4,6 +4,21 @@ let port = ref None
+  (* Map sample_rate to the string statsd expects *)
+  let sample_rate =
+    if sample_rate >= 1.0 then
+      ""  (* not sampling *)
```
This is confusing. Why does >= 1 mean we don't send anything?

bradlangel replied:
NOTE: This logic is copied from before with no changes. 

Not sampling here just means that we are sending all data along. I.e. we have a sampling of 100% or 1.0. This logic seems faulty to me since sample-rate is a float and a negative sampling doesn't make sense. Nor does anything greater than 1.0. However, the error handling in this library is pretty poor, imo so its not accounted for. The way that dogstatsd handles this is as follows: https://github.com/DataDog/datadogpy/blob/eff9a2a03f92979fa17011df5b63a926a19802e2/datadog/dogstatsd/base.py#L279

brendanlong reviewed (approved):
I suspect the helper functions belong in a different file but I think this looks ok.

bradlangel committed:
Exposes statsd format and send functions

Motivation for this commit is to give the user access to statsd message
formatting functions and a low level send function in order allow the
user to extend the statsd protocol. This also moves the serializing
the data into a string out of the send function.

kevinqiu commented on statsd-client/statsd_client_core.ml:
```diff
@@ -51,83 +65,77 @@ module Make (U : IO) :
+      send_with_ipaddr_and_port ipaddr port (fmt data)
+
+  let fmt_data ?(sample_rate = 1.0) data =
+    if sample_rate >= 1.0 || sample_rate >= Random.float 1.0
```
I remember you told me in person, but what was the reasoning for moving this logic into the `fmt_data` fn from the `send` fn?

bradlangel replied:
* the send function isn't dependent upon the sample rate like it was prior to this revision
* check that our ip and port are set prior to formatting our data
I've refactored the code since we chatted and as a result, there really isn't a compelling reason anymore. This might be something that can be refactored out as its no longer that useful and at best is a small optimization.
Here is an example of this being used in dogstatsd: https://github.com/arenadotio/ocaml-statsd-client/pull/4/commits/5ab42c1806ecc69c4b8996239d3d0acdd23b1546#diff-ee22f142e5f6184713e7201d0eccc1a0R147

I don't feel strongly about it but do you think I should refactor the code to be such that the signature of `send` is:
```
val send: string list -> unit _r
```
Where `string list` is the formatted data?

kevinqiu replied:
Haha, I was wondering if we needed to pass `~fmt` to `send` but didn't know how we were planning to use it in the future.
